### ISSUE TITLE
Notifications about uncaught exceptions that happen during runtime

### ISSUE THREAD

**Issue by Bios-Marcel** [TradeWars/Economy#2]
The JVM offers registering an `UncaughtExceptionHandler`, which allows us to react to `exceptions` that haven't been caught by any `try-catch`. Inside of the `UncaughtExceptionHandler` we have the option to inform someone about the `Exception`. As we don't have any GUI on which you'd notice the occurrence of an `Exception`, we could send an E-Mail, Discord-Notification or automatically create an issue on GitHub.

Remove the TODO in [UncaughtExceptionHandler](https://github.com/TradeWars/Economy/blob/develop/src/main/kotlin/org/tradewars/economy/basic/UncaughtExceptionHandler.kt) after finishing this issue.

**Comment by ghost** [comment]
The only problem I can think of is that it might spam. For example if the exception occurs inside of a loop or something else repetitive. I guess this can be prevented though.

Email seems a bit private, this is an open-source project after all. Besides, we'd have to constantly check for new emails.
Discord is used by all of us (you, Southclaws and I), and we'd get notified as soon as the issue occurs. However if our group of contributors grow, this might become an issue as some of them might not use or want to use Discord.
I think issues should only be used by actual people, and not get filled up with error messages.

Despite the issue I mentioned with Discord, I still believe that's the best option for now.
I'm up for experimenting with this.

**Comment by Bios-Marcel** [comment]
So we can theoretically remember what notifications we have sent already and if an `Exception` is equal to one that we have sent a notification about in the past, we will simply ignore it. On the other hand, it might be interesting to see how often given `Exception` occurs. 

Maybe we should send the notification only once, but persist information about occurred `Exception`s. I am already logging every `Exception` that occurs, but I don't think that log files are very readable, it might be better saving the information structured and write a little console application that pretty prints the contained information.

Doesn't matter if it sounds like over-engineering, as long as it is fun and somewhat of use.

**Comment by ghost** [comment]
That's what I was thinking. However, in some situations each exception might be slightly different on every iteration.

What do you think about storing them in JSON?

**Comment by Bios-Marcel** [comment]
Yeah, that might be an option. Let's wait for @Southclaws to throw his opinion in.

**Comment by Southclaws** [comment]
If exception output can be generalised with something like [this](https://stackoverflow.com/questions/22271099/convert-exception-to-json) then we can handle the alerts outside of the application - this would be my preferred method because:

1. it removes the responbility from the application
2. it re-uses code
3. it removes the need to write the same type of alerting system into all codebases

But I still think it could be useful having it built in too... what are your thoughts on this? Because it ties in quite a lot with #3

**Comment by Southclaws** [comment]
And yes, Email is not good because I don't consider email an instant transaction, it's something I (and many others) check when *they* want to rather than allow email to control *them* by using notifications. Discord is the best place for this now, I don't expect any future contributors to not use discord (primarily because I will probably spread the word of this project only via discord anyway...)

**Comment by Bios-Marcel** [comment]
Yeah, the `Exception` to `JSON` thing wouldn't be a problem, convert and append to file. We could then have some kind of application that watches the file for changes.

**Comment by Southclaws** [comment]
no files! read #3

**Comment by Bios-Marcel** [comment]
First implementation mentioned at: https://github.com/TradeWars/Economy/issues/3#issuecomment-406899410
